vignettes/python_packages.Rmd

```diff
@@ -1,13 +1,157 @@
 ---
 title: "Installing Python Packages"
-output: 
+output:
   rmarkdown::html_vignette
+editor_options:
+  markdown:
+    wrap: 80
 vignette: >
   %\VignetteIndexEntry{Installing Python Packages}
   %\VignetteEngine{knitr::rmarkdown}
   %\VignetteEncoding{UTF-8}
 ---
 
+## Declaring Python Requirements with `py_require()`
+
+`py_require()` is the recommended way to declare Python dependencies. A
+`py_require()` call is similar to a `library()` call for Python packages.
+
+`py_require()`:
+
+-   Should appear at the start of an R script.
+-   Declares the Python dependencies needed by the script.
+-   Is limited in scope to the current R session.
+-   Has no effect if called multiple times.
+
+Additionally, R packages may call `py_require()` with their own dependencies
+when loaded. Requirements declared by an R package are combined with those
+declared by users.
+
+It is generally recommended, though not required, to declare all Python
+dependencies before reticulate initializes Python. The simplest way to ensure
+this is by grouping all `py_require()` calls at the start of a script, alongside
+`library()` calls.
+
+Usage example:
+
+``` r
+library(reticulate)
+py_require("jax")
+
+jax <- import("jax") # <-- initializes Python
+```
+
+### Unsatisfiable requirements.
+
+It's possible to inadvertently declare an unsatisfiable set of Python
+requirements. This can happen for a variety of reasons, most often due to:
+
+-   Incompatible requirements declared by different R packages.
+-   A recently published Python package version that introduces conflicts.
+
+To troubleshoot:
+
+1.  Figure out what the conflicting requirements are and where they are coming
+    from.
+2.  Update the requirements to resolve conflicts.
+
+To help with step 1, calling `py_require()` with no arguments will print out a
+manifest of the current requirements, as well as a table showing what
+requirements were declared by different R packages.
+
+Additionally, attempting to initialize Python will print out a detailed error
+message.
+
+For example:
+
+``` r
+library(reticulate)
+py_require("numpy>2")
+py_require("numpy<2")
+import("numpy")
+```
+
+```         
+  × No solution found when resolving `--with` dependencies:
+  ╰─▶ Because you require numpy>2 and numpy<2, we can conclude that your
+      requirements are unsatisfiable.
+uv error code: 1
+-- Current requirements -------------------------------------------------
+ Python:   3.11.11 (reticulate default)
+ Packages: numpy, numpy>2, numpy<2
+-------------------------------------------------------------------------
+Error in uv_get_or_create_env() :
+  Call `py_require()` to remove or replace conflicting requirements.
+```
+
+You can call `py_require()` again with `action = "remove"` or `action = "set"`
+to update the requirements to make them satisfiable. Options for resolving are:
```

Comment:
There is a colon at the end here, but No 1 doesn't start until after the next paragraph, not sure if that was on purpose